app/services/ProcessCsvService.scala

```diff
@@ -0,0 +1,186 @@
+    _.flatMapConcat(extractEntityData)
+      .via(CsvParsing.lineScanner())
+      .via(Flow.fromFunction(Right(_)))
+      .recover {
+        case e => Left(e)
+      }
+
+  def processFiles(callback: Option[UpscanCsvFilesCallbackList], scheme: String, source: String => Source[HttpResponse, _])(
+    implicit request: RequestWithOptionalEmpRef[AnyContent], hc: HeaderCarrier, messages: Messages
+  ): List[Future[Either[Throwable, Boolean]]] =
+    callback.get.files map { file =>
+
+      val successUpload = file.uploadStatus.asInstanceOf[UploadedSuccessfully]
+
+      val validatorFuture: Future[Either[Throwable, DataValidator]] = Source(List(successUpload.name))
+        .via(Flow.fromFunction(checkFileType(_)(messages)))
+        .via(eitherFromFunction(dataGenerator.getSheetCsv(_, scheme)(messages)))
+        .via(eitherFromFunction(dataGenerator.identifyAndDefineSheetEither(_)(hc, request, messages)))
+        .via(eitherFromFunction(dataGenerator.setValidatorCsv(_)(hc, request, messages)))
+        .runWith(Sink.head)
+
+      validatorFuture.flatMap {
+        _.fold(
+          throwable => Future.successful(Left(throwable)),
+          validator => {
+            val futureListOfErrors: Future[Seq[Either[Throwable, RowValidationResults]]] = extractBodyOfRequest(source(successUpload.downloadUrl))
+              .via(eitherFromFunction(processRow(_, successUpload.name, validator)))
+              .runWith(Sink.seq[Either[Throwable, RowValidationResults]])
+
+            futureListOfErrors.map {
+              getRowsWithNumbers(_, successUpload.name)(messages)
+            }.flatMap{
+              case Right(errorsFromRow) => checkValidityOfRows(errorsFromRow, successUpload.name, file)
+              case Left(exception) => Future(Left(exception))
+            }
+          }
+        )
+      }
+    }
+
+  def processRow(rowBytes: List[ByteString], sheetName: String, validator: DataValidator): Either[Throwable, RowValidationResults] = {
+    val rowStrings = rowBytes.map(byteString => byteString.utf8String)
+    val parsedRow = parserUtil.formatDataToValidate(rowStrings, sheetName)
+    val rowIsEmpty = parserUtil.rowIsEmpty(rowStrings)
+
+    Try {
+      validator.validateRow(Row(0, getCells(parsedRow, 0)))
+    } match {
+      case Failure(e) =>
+        Logger.warn(e.toString)
+        Left(e)
+      case Success(list) => Right(RowValidationResults(list.getOrElse(List.empty), rowIsEmpty))
+    }
+  }
+
+  def checkFileType(name: String)(implicit messages: Messages): Either[Throwable, String] = if (name.endsWith(".csv")) {
+    Right(FilenameUtils.removeExtension(name))
+  } else {
+    Left(ERSFileProcessingException(
+      Messages("ers_check_csv_file.file_type_error", name)(messages),
+      Messages("ers_check_csv_file.file_type_error", name)(messages)))
+  }
+
+@tailrec
+private[services] final def processDisplayedErrors(n: Int, list: Seq[(List[ValidationError], Int)]): Seq[(List[ValidationError], Int)] = {
```

Comment:
We've discussed but putting a comment here for visibility. This needs more tests currently.
Easier readability through not using `n` and `._1` where possible.